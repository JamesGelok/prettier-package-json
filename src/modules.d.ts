declare module 'sort-order' {
  type SortFn<T> = (a: T, b: T) => 1 | 0 | -1;

  function orderBy<T>(...sortFns: SortFn<T>[]): SortFn<T>;

  export = orderBy;
}
