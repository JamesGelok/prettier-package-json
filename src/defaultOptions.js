module.exports = Object.freeze({
  tabWidth: 2,
  keyOrder: [
    'name',
    'description',
    'author',
    'license',
    'repository',
    'bugs',
    'version',
    'main',
    'bin',
    'scripts',
    'dependencies',
    'peerDependencies',
    'devDependencies',
    'optionalDependencies',
    'keywords'
  ]
});
