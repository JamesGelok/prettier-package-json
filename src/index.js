
const defaults = require('./defaultOptions');
const sortObject = require('sort-object-keys');

function format(packageJson, opts) {
  const options = Object.assign({}, defaults, opts);
  return JSON.stringify(sortObject(packageJson, options.keyOrder), null, options.tabWidth);
}

module.exports = {
  format
};
