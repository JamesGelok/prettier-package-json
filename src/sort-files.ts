/**
 * Sort and filter files field
 *
 * More info:
 *   https://docs.npmjs.com/files/package.json#files
 */

import minimatch from 'minimatch';
import { PackageJson } from './types';

type FilterFn<T> = (...args: T[]) => boolean;

const not = <T>(filterFn: FilterFn<T>) => (arg: T) => !filterFn(arg);
const or = <T>(...filterFns: FilterFn<T>[]) => (arg: T) => filterFns.some((fn) => fn(arg));

const ALWAYS_INCLUDED = [
  /^package.json$/,
  /^README.*/i,
  /^CHANGE(S|LOG).*/i,
  /^HISTORY.*/i,
  /^LICEN(C|S)E.*/i,
  /^NOTICE.*/i
]
  .map((regex) => (filepath: string) => regex.test(filepath))
  .reduce((a, b) => or(a, b));

const ALWAYS_EXCLUDED = [
  '.git',
  'CVS',
  '.svn',
  '.hg',
  '.lock-wscript',
  '.wafpickle-N',
  '.*.swp',
  '.DS_Store',
  '._*',
  'npm-debug.log',
  '.npmrc',
  'node_modules',
  'config.gypi',
  '*.orig',
  'package-lock.json'
]
  .map((glob) => (minimatch.filter(glob) as any) as FilterFn<string>)
  .reduce((a, b) => or(a, b));

export default function sortFiles(packageJson: PackageJson): { files?: PackageJson['files'] } {
  const { files = [], main } = packageJson;

  const isPackageMain = (filepath: string) => filepath === main;
  const ignored = or(ALWAYS_INCLUDED, ALWAYS_EXCLUDED, isPackageMain);

  const sortedAndFilteredFiles = files.filter(not(ignored)).sort();

  return sortedAndFilteredFiles.length > 0 ? { files: sortedAndFilteredFiles } : {};
}
