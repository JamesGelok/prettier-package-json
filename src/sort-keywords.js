module.exports = function sortKeywords(keywords = []) {
  return keywords.length === 0 ? {} : { keywords: keywords.sort() };
};
