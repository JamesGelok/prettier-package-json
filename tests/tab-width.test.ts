import { format } from '../src/index';

test('It indents using 2 spaces (default)', () => {
  const json = {
    name: 'Test',
    description: 'Description',
    version: '0.0.0'
  };

  expect(format(json)).toMatchSnapshot();
});

test('It indents using the specified tab width', () => {
  const json = {
    name: 'Test',
    description: 'Description',
    version: '0.0.0'
  };

  expect(format(json, { tabWidth: 4 })).toMatchSnapshot();
});
