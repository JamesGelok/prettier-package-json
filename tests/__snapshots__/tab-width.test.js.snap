// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It indents using a default tab width 1`] = `
"{
  \\"name\\": \\"Test\\",
  \\"description\\": \\"Description\\",
  \\"version\\": \\"0.0.0\\"
}"
`;

exports[`It indents using the specified tab width 1`] = `
"{
    \\"name\\": \\"Test\\",
    \\"description\\": \\"Description\\",
    \\"version\\": \\"0.0.0\\"
}"
`;
